Add functional test test_txid_inv_delay

ariard opened the pull request:
This is a simple functional test to increase coverage of #19988, checking that txid announcements from txid-relay peers are delayed by TXID_RELAY_DELAY, assuming we have at least another wtxid-relay peer.

You can verify new test with the following diff :

```
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index f14db379f..2a2805df5 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -773,7 +773,7 @@ void PeerManager::AddTxAnnouncement(const CNode& node, const GenTxid& gtxid, std
     auto delay = std::chrono::microseconds{0};
     const bool preferred = state->fPreferredDownload;
     if (!preferred) delay += NONPREF_PEER_TX_DELAY;
-    if (!gtxid.IsWtxid() && g_wtxid_relay_peers > 0) delay += TXID_RELAY_DELAY;
+    //if (!gtxid.IsWtxid() && g_wtxid_relay_peers > 0) delay += TXID_RELAY_DELAY;
     const bool overloaded = !node.HasPermission(PF_RELAY) &&
         m_txrequest.CountInFlight(nodeid) >= MAX_PEER_TX_REQUEST_IN_FLIGHT;
     if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
```

sipa commented on test/functional/p2p_segwit.py:
```diff
@@ -148,7 +148,7 @@ def test_witness_block(node, p2p, block, accepted, with_witness=True, reason=Non
 class TestP2PConn(P2PInterface):
     def __init__(self, wtxidrelay=False):
-        super().__init__()
+        super().__init__(wtxidrelay=wtxidrelay)
```
Can you make the changes to the test framework in a separate commit?

ariard replied:
Sure, I did it at first before to squash. Split in two.

sipa commented on test/functional/p2p_tx_download.py:
```diff
@@ -216,6 +216,24 @@ def test_preferred_inv(self):
+        peer.sync_with_ping()
+        with p2p_lock:
+            assert_equal(peer.tx_getdata_count, 0)
+        self.nodes[0].setmocktime(mock_time + TXID_RELAY_DELAY)
```
Wouldn't it be better to start mocking before the inv is sent? So that local timing doesn't matter?

ariard replied:
I think that's already done L223, we set mocktime for nodes 0 before it's even connected to the first p2p connection. Or do you have a more constraining definition of mocktime in mind ?

sipa replied:
Oh, of course!

sipa reviewed (commented):
Concept ACK

Linter error: test/functional/p2p_segwit.py:14:1: F401 'test_framework.messages.msg_verack' imported but unused

Commit:
test: Makes wtxidrelay support a generic P2PInterface option

Its usage is extended beyond p2p_segwit.py in next commit.

Commit:
Add functional test test_txid_inv_delay

Add a simple functional test to cover TXID_RELAY_DELAY as applied
as a TxRequestTracker parameter in AddTxAnnoucement.

ariard commented:
Thanks, updated at f0fe840. See if this [comment](https://github.com/bitcoin/bitcoin/pull/20171#discussion_r506750652) needs further modification.

naumenkogs commented on test/functional/p2p_tx_download.py:
```diff
@@ -216,6 +216,24 @@ def test_preferred_inv(self):
+        self.nodes[0].setmocktime(mock_time + TXID_RELAY_DELAY)
+        peer.wait_until(lambda: peer.tx_getdata_count >= 1, timeout=1)
+        with p2p_lock:
+            assert_equal(peer.tx_getdata_count, 1)
```
`wait until` right above wouldn't allow this line to happen if `tx_getdata_count == 0`. So we're checking it's NOT more than 1? Is it a worthy check?...

ariard replied:
Right, in fact it sounds other new functional tests added with #19988 have a redundant p2p lock tacking so added a new commit to remove this oversight. Unless @MarcoFalke they serve something else ?

naumenkogs commented on test/functional/p2p_tx_download.py:
```diff
@@ -216,6 +216,24 @@ def test_preferred_inv(self):
+        mock_time = int(time.time() + 1)
+        self.nodes[0].setmocktime(mock_time)
+        peer = self.nodes[0].add_p2p_connection(TestP2PConn(wtxidrelay=False))
+        # Add a second wtxid-relay connection otherwise TXID_RELAY_DELAY is waived in
```
perhaps also worth testing?

ariard replied:
Added a test mutation.

naumenkogs commented on test/functional/p2p_tx_download.py:
```diff
@@ -216,6 +216,24 @@ def test_preferred_inv(self):
+        peer.sync_with_ping()
+        with p2p_lock:
+            assert_equal(peer.tx_getdata_count, 0)
+        self.nodes[0].setmocktime(mock_time + TXID_RELAY_DELAY)
```
Shouldn't `NONPREF_PEER_TX_DELAY` also be applied here?

Anyway, I'd first jump in `NONPREF_PEER_TX_DELAY` to see that it's not enough time passed, and then jump `TXID_RELAY_DELAY` again to see that enough time passed? I think that's what is supposed to happen..

ariard replied:
The positive case was already tested in `test_preferred_inv` but added a mutation to cover the negative one, i.e applying `NONPREF_PEER_TX_DELAY` on non-preferred peers.

naumenkogs commented:
Concept ACK

glozow reviewed (commented):
Concept ACK

MarcoFalke commented:
ACK f0fe840ace0d317dcde48d1f21554f884accbbf9

Commit:
Add mutation for functional test test_preferred_inv

Add a booelan arg to test_preferred_inv to cover NONPREF_PEER_TX_DELAY
as applied as a TxRequestTracker parameter in AddTxAnnouncement.

Commit:
Remove redundant p2p lock tacking for tx download functional tests

New functional test coverage of tx download was added by #19988,
but `with p2p_lock` is redundant for some tests with `wait_until`
test helper, already guaranteeing test lock tacking.

ariard commented on test/functional/p2p_tx_download.py:
```diff
@@ -216,6 +216,24 @@ def test_preferred_inv(self):
+        peer.sync_with_ping()
+        with p2p_lock:
+            assert_equal(peer.tx_getdata_count, 0)
+        self.nodes[0].setmocktime(mock_time + TXID_RELAY_DELAY)
```
I don't think you can achieve this with current p2p framework, we can write to the mocktime but can we read it from it ? A quick look, I don't find test doing it so not sure that's a feature of our current test framework. Note that was already the way done by legacy test. But lmk if there is way to do so I don't see

glozow replied:
> I don't find test doing it so not sure that's a feature of our current test framework

Yeah, I haven't seen an example either. The way I'd do it is just use `sleep()` instead of setmocktime... 2 seconds isn't too bad when the tests are running in parallel, but fosho not ideal.

ariard commented:
Thanks @naumenkogs, @glozow for reviews, updated addressing your comments at c55ce67

jonatack commented on test/functional/p2p_tx_download.py:
```diff
@@ -47,6 +47,7 @@ def on_getdata(self, message):
 OVERLOADED_PEER_DELAY = 2 # seconds
 MAX_GETDATA_IN_FLIGHT = 100
 MAX_PEER_TX_ANNOUNCEMENTS = 5000
+NONPREF_PEER_TX_DELAY=2
```
```suggestion
NONPREF_PEER_TX_DELAY = 2
```

jonatack commented on test/functional/p2p_tx_download.py:
```diff
@@ -204,17 +201,45 @@ def test_notfound_fallback(self):
+            if glob_wtxid:
+                    assert_equal(peer.tx_getdata_count, 0)
+            else:
+                    assert_equal(peer.tx_getdata_count, 1)
```
Tabs 4 spaces instead of 8, but you could also simplify here with
```diff
-            if glob_wtxid:
-                    assert_equal(peer.tx_getdata_count, 0)
-            else:
-                    assert_equal(peer.tx_getdata_count, 1)
+            assert_equal(peer.tx_getdata_count, 0 if glob_wtxid else 1)
```

ariard replied:
Less is better.

jonatack commented on test/functional/test_framework/p2p.py:
```diff
@@ -394,7 +397,7 @@ def on_verack(self, message):
     def on_version(self, message):
         assert message.nVersion >= MIN_VERSION_SUPPORTED, "Version {} received. Test framework only supports versions greater than {}".format(message.nVersion, MIN_VERSION_SUPPORTED)
-        if message.nVersion >= 70016:
+        if message.nVersion >= 70016 and self.wtxidrelay :
```
```suggestion
        if message.nVersion >= 70016 and self.wtxidrelay:
```

BARMOLEY007 replied:
Gut

jonatack reviewed (commented):
Approach ACK. Verified the new test does fail per the PR description.

MarcoFalke commented:
How much difference is this going to make in practice? For a node without incoming connections it seems unlikely to have peers that don't support the `feefilter` message. For a node with incoming connections, minor traffic waste doesn't seem to be a concern.

ariard commented:
Thanks @jonatack. I think I took everything at bc4a230

ariard commented:
@naumenkogs @jonatack @glozow @MarcoFalke if you have few minutes to review again this PR, it hasn't changed and I think all previous considerations have been addressed.

laanwj commented:
ACK bc4a23008762702ffcd6868bcdb8fe2a732640ba
Thanks for adding a test for this.
